src/tools/add_entity.cc

```diff
@@ -203,63 +170,27 @@ struct add_entity_tool : tool {
 
     void get_description(char *str) override {
         auto name = entity_names[entity_name_index];
-        const char *rotate = nullptr;
-        switch (rotate_mode) {
-            case RotateMode::AxisAligned: {
-                rotate = "Axis Aligned";
-                break;
-            }
-            case RotateMode::Stepped45: {
-                rotate = "Stepped @ 45";
-                break;
-            }
-            case RotateMode::Stepped15: {
-                rotate = "Stepped @ 15";
-                break;
-            }
-        }
 
-        const char *place = nullptr;
-        switch (place_mode) {
-            case PlaceMode::BlockSnapped: {
-                place = "Block Snapped";
-                break;
-            }
-            case PlaceMode::HalfBlockSnapped: {
-                place = "Half Block Snapped";
-                break;
-            }
-            case PlaceMode::QuarterBlockSnapped: {
-                place = "Quarter Block Snapped";
-                break;
-            }
-            case PlaceMode::EighthBlockSnapped: {
-                place = "Eighth Block Snapped";
-                break;
-            }
-            case PlaceMode::FreeForm: {
-                place = "Free Form";
-                break;
-            }
-        }
-        sprintf(str, "Place %s \nRotating %s \nPlacing %s", name.c_str(), rotate, place);
+        sprintf(str, "Place %s", name.c_str());
     }
 
     glm::mat4 get_place_matrix(const raycast_info *rc, unsigned int index) const {
         glm::mat4 m;
         auto rot_axis = glm::vec3{surface_index_to_normal(surface_zp)};
 
+        auto place = entity_stubs[entity_names[entity_name_index]].get_component<placeable_component_stub>();
+
         float step = 1;
-        switch (place_mode) {
-            case PlaceMode::EighthBlockSnapped: {
+        switch (place->place) {
+            case placement::eighth_block_snapped: {
                 step *= 2;
                 // falls through
             }
-            case PlaceMode::QuarterBlockSnapped: {
+            case placement::quarter_block_snapped: {
                 step *= 2;
                 // falls through
             }
-            case PlaceMode::HalfBlockSnapped: {
+            case placement::half_block_snapped: {
                 step *= 2;
                 auto pos = rc->intersection;
                 m = mat_rotate_mesh(pos, rc->n);
```

Comment:
This rounding logic below should just round all 3 axes. It's simpler, and it tidies up any small inaccuracy in `rc->intersection`

Reply:
I'm too tired to think about what you mean

Reply:
You're carefully snapping the two tangent axes but not the normal axis. It should be safe to snap all 3.

Reply:
(Not new in this change; just seems like a nice simplification)

Reply:
Oh. That seems like it might generate the wrong results? I guess it depends on how far `rc->intersection` can ever be?

Reply:
If it's ever > 1/16 of a block I've failed horribly

Reply:
Is this what you mean?
```
                if (n.x != 0) {
                    p.x = std::round(p.x);
                    p.y = std::round(p.y * step) / step;
                    p.z = std::round(p.z * step) / step;
                } else if (n.y != 0) {
                    p.x = std::round(p.x * step) / step;
                    p.y = std::round(p.y);
                    p.z = std::round(p.z * step) / step;
                } else if (n.z != 0) {
                    p.x = std::round(p.x * step) / step;
                    p.y = std::round(p.y * step) / step;
                    p.z = std::round(p.z);
                }```

Reply:
if so, pushed already.

Reply:
I'm saying this should be fine:
```
p.x = std::round(p.x * step) / step;
p.y = std::round(p.y * step) / step;
p.z = std::round(p.z * step) / step;
```

If we're ever off the surface by more than `step/2` then we've got bigger issues.

Reply:
done